firestore/integration_test_internal/src/jni/arena_ref_test.cc

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "firestore/src/jni/arena_ref.h"
+#include "firestore/src/android/firestore_android.h"
+#include "firestore/src/common/make_unique.h"
+#include "firestore/src/jni/loader.h"
+#include "firestore/src/jni/long.h"
+
+#include "firestore_integration_test.h"
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace jni {
+namespace {
+
+constexpr char kException[] = "java/lang/Exception";
+Method<String> kConstructor("<init>", "(Ljava/lang/String;)V");
+
+class ArenaRefTestAndroid : public FirestoreIntegrationTest {
+ public:
+  ArenaRefTestAndroid() : loader_(app()), env_(make_unique<Env>(GetEnv())) {
+    loader_.LoadClass(kException);
+    loader_.Load(kConstructor);
+  }
+
+  ~ArenaRefTestAndroid() override {
+    // Ensure that after the test is done that any pending exception is cleared
+    // to prevent spurious errors in the teardown of FirestoreIntegrationTest.
+    env_->ExceptionClear();
+  }
+
+  Env& env() const { return *env_; }
+
+  void throwException() {
+    Local<Class> clazz = env().FindClass(kException);
+    jmethodID ctor =
+        env().GetMethodId(clazz, "<init>", "(Ljava/lang/String;)V");
+
+    Local<String> message = env().NewStringUtf("Testing throw");
+    Local<Throwable> exception = env().New<Throwable>(clazz, ctor, message);
+    // After throwing, use EXPECT rather than ASSERT to ensure that the
+    // exception is cleared.
+    env().Throw(exception);
+    EXPECT_FALSE(env().ok());
+  }
+
+  void clearExceptionOccurred() {
+    Local<Throwable> thrown = env().ClearExceptionOccurred();
+    EXPECT_EQ(thrown.GetMessage(env()), "Testing throw");
+  }
+
+ private:
+  // Env is declared as having a `noexcept(false)` destructor, which causes the
+  // compiler to propagate this into `EnvTest`'s destructor, but this conflicts
+  // with the declaration of the parent class. Holding `Env` with a unique
+  // pointer sidesteps this restriction.
+  std::unique_ptr<Env> env_;
+  Loader loader_;
```

Comment:
Don't hold an Env in an instance variable. Instead, just do `Env env` when you need one. And methods of this class that need one should take one as a non-const reference.